serveradmin/serverdb/views.py

```diff
@@ -1,181 +1,145 @@
-        'application': application,
+        'from': f_from,
+        'until': f_until,
+        'hostname': f_hostname,
+        'object_id': f_object_id,
+        'user_or_app': f_user_or_app,
     })
-    return TemplateResponse(request, 'serverdb/changes.html', context)
 
 
 @login_required
 def history(request):
     object_id = request.GET.get('object_id')
-    commit_id = request.GET.get('commit_id')
-    search_string = request.GET.get('search_string')
-    exclude_string = request.GET.get('exclude_string')
-
-    if not object_id:
-        raise Http404
-
-    # TODO: Change data model so that server_id is moved to ChangeCommit.
-    #       To me this would make most sense. If that is not possible (why?!)
-    #       write a proper (custom) query that performs.
-    where = {'server_id': object_id}
-    adds = ChangeAdd.objects.filter(**where)
-    updates = ChangeUpdate.objects.filter(**where)
-    deletes = ChangeDelete.objects.filter(**where)
-
-    # @TODO Transform to PostgreSQL JSON and put a index here
-    if search_string:
-        adds = adds.filter(attributes_json__contains=search_string)
-        updates = updates.filter(updates_json__contains=search_string)
-        deletes = deletes.filter(attributes_json__contains=search_string)
-
-    if exclude_string:
-        adds = adds.exclude(attributes_json__contains=exclude_string)
-        updates = updates.exclude(updates_json__contains=exclude_string)
-        deletes = deletes.exclude(attributes_json__contains=exclude_string)
 
+    if not Server.objects.filter(server_id=object_id).exists():
+        messages.error(
+            request,
+            f'Object with id {object_id} does not or no longer exist.')
+        return redirect(reverse('serverdb_changes'))
+
+    server = get_object_or_404(Server, server_id=object_id)
+    obj_history = Change.objects.filter(object_id=object_id)
+
+    commit_id = request.GET.get('commit_id')
     if commit_id:
-        adds = adds.filter(commit__pk=commit_id)
-        updates = updates.filter(commit__pk=commit_id)
-        deletes = deletes.filter(commit__pk=commit_id)
+        obj_history = obj_history.filter(commit_id=commit_id)
 
-    related = ('commit__app', 'commit__user')
-    adds = adds.select_related(*related)
-    updates = updates.select_related(*related)
-    deletes = deletes.select_related(*related)
+    attribute_filter = request.GET.get('attribute_filter')
+    if attribute_filter:
+        obj_history = obj_history.filter(change_json__has_key=attribute_filter)
 
-    change_list = ([('add', obj) for obj in adds] +
-                   [('update', obj) for obj in updates] +
-                   [('delete', obj) for obj in deletes])
-    change_list.sort(key=lambda entry: entry[1].commit.change_on, reverse=True)
+    obj_history = obj_history.order_by('-commit_id')
+    obj_history = obj_history.select_related('commit__app', 'commit__user')
+
+    page = request.GET.get('page', 1)
+    pager = Paginator(obj_history, 25)
```

Comment:
Might be a matter of taste but I find `25` a little bit too few. I would vote for `100` or something like that

Reply:
25 is what works on most modern screen sizes in my experience.

So you would be fine with scrolling ?

Reply:
Ah ok you mean because the page links are at the bottom 🤔 Well in that case I am also fine with 25. I was just thinking about object creates or the initial puppetruns where a lot of changes are made at once and they would not fit in one page

Reply:
Lets keep 25 for now. It would be easy to adjust it later.